src/knot_thing_main.h

```diff
@@ -7,74 +7,86 @@
  *
  */
 
-#ifndef __KORE_SENSOR_H__
-#define __KORE_SENSOR_H__
-
-struct sensor_integer {
-	const char	*name;		/* Application defined sensor name */
-	uint8_t		sensor_id;	/* Application defined id	*/
-	uint16_t	type_id;	/* See KNOT_TYPE_ID_xxx		*/
-	uint8_t		unit;		/* See KNOT_UNIT_xxx		*/
-
-	/* Read data from sensor */
-	int (*read)(int32_t *val, int32_t *multiplier);
-
-	/* Reports new data to sensor */
-	int (*write)(int32_t val, int32_t multiplier);
-};
-
-struct sensor_float {
-	const char	*name;		/* Application defined sensor name */
-	uint8_t		sensor_id;	/* Application defined id	*/
-	uint16_t	type_id;	/* See KNOT_TYPE_ID_xxx		*/
-	uint8_t		unit;		/* See KNOT_UNIT_xxx		*/
-
-	/* Read data from sensor */
-	int (*read)(int32_t *val, int32_t *multiplier, int32_t *valdec);
-
-	/* Reports new data to sensor */
-	int (*write)(int32_t val, int32_t multiplier, int32_t valdec);
-};
-
-struct sensor_bool {
-	const char	*name;		/* Application defined sensor name */
-	uint8_t		sensor_id;	/* Application defined id	*/
-	uint16_t	type_id;	/* See KNOT_TYPE_ID_xxx		*/
-	uint8_t		unit;		/* See KNOT_UNIT_xxx		*/
-
-	/* Read data from sensor */
-	int (*read)(int8_t *val);
-
-	/* Reports new data to sensor */
-	int (*write)(int8_t val);
-};
-
-struct sensor_raw {
-	const char	*name;		/* Application defined sensor name */
-	uint8_t		sensor_id;	/* Application defined id	*/
-	uint16_t	type_id;	/* See KNOT_TYPE_ID_xxx		*/
-	uint8_t		unit;		/* See KNOT_UNIT_xxx		*/
-
-	/* Read data from sensor */
-	int (*read)(void *buffer, int8_t *buffer_len);
-
-	/* Reports new data to sensor */
-	int (*write)(void *buffer, int8_t buffer_len);
-};
-
-/* KNOT core (kore) initialization functions and polling */
-int8_t kore_init(void);
-void kore_exit(void);
-
-int8_t kore_run(void);
+#ifndef __KNOT_THING_MAIN_H__
+#define __KNOT_THING_MAIN_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "knot_protocol.h"
+
+typedef int (*intDataFunction)		(int32_t *val, int32_t *multiplier);
+typedef int (*floatDataFunction)	(int32_t *val_int, uint32_t *val_dec, int32_t *multiplier);
```

Comment:
It is important to declare as a pointer because this function definition will be used for read and write operations.